lib/Conversion/ImportVerilog/Statement.cpp

```diff
@@ -19,121 +19,311 @@
+  }
 
-// It can handle the statements like case, conditional(if), for loop, and etc.
-LogicalResult
-Context::convertStatement(const slang::ast::Statement *statement) {
-  auto loc = convertLocation(statement->sourceRange.start());
-  switch (statement->kind) {
-  case slang::ast::StatementKind::Empty:
+  LogicalResult visit(const slang::ast::VariableDeclStatement &) {
+    // TODO: not sure
     return success();
-  case slang::ast::StatementKind::List:
-    for (auto *stmt : statement->as<slang::ast::StatementList>().list)
-      if (failed(convertStatement(stmt)))
+  }
+
+  LogicalResult visit(const slang::ast::ExpressionStatement &exprStmt) {
+    return success(context.convertExpression(
+        exprStmt.as<slang::ast::ExpressionStatement>().expr));
+  }
+
+  LogicalResult visit(const slang::ast::StatementList &listStmt) {
+    for (auto *stmt : listStmt.list) {
+      auto succeeded = (*stmt).visit(*this);
+      if (succeeded.failed())
+        return succeeded;
+    }
+    return success();
+  }
+
+  LogicalResult visit(const slang::ast::BlockStatement &blockStmt) {
+    Context::SymbolTableScopeT varScope(context.varSymbolTable);
+    return blockStmt.body.visit(*this);
+  }
+
+  LogicalResult visit(const slang::ast::EmptyStatement &emptyStmt) {
+    return success();
+  }
+
+  LogicalResult visit(const slang::ast::TimedStatement &timeStmt) {
+    if (failed(context.visitTimingControl(
+            &timeStmt.as<slang::ast::TimedStatement>().timing)))
+      return failure();
+    if (failed(timeStmt.stmt.visit(*this)))
+      return failure();
+
+    return success();
+  }
+
+  /// Handle Loop
+  LogicalResult visit(const slang::ast::ForLoopStatement &forStmt) {
+    // reuse scf::whileOp to rewrite ForLoop
+    // ------------
+    // for (init_stmt; cond_expr; step_stmt) begin
+    //   statements
+    // end
+    // -------------
+    // init_stmt;
+    // while (cond_expr) {
+    //   body;
+    //   step_stmt;
+    // }
+    // -------------
+    mlir::SmallVector<mlir::Type> types;
+    auto type = context.convertType(*forStmt.stopExpr->type, loc);
+
+    auto whileOp = builder.create<mlir::scf::WhileOp>(
+        loc, types, mlir::SmallVector<Value, 0>{});
+    OpBuilder::InsertionGuard guard(builder);
+
+    // The before-region of the WhileOp.
+    Block *before = builder.createBlock(&whileOp.getBefore());
+    builder.setInsertionPointToEnd(before);
+    Value cond = context.convertExpression(*forStmt.stopExpr);
+    if (!cond)
+      return failure();
+
+    // The numeric value of the whileOp condition expression is tested for being
+    // zero. And while (cond_expression) is equivalent to
+    // while (cond_expression != 0). So the following code is for handling
+    // `while (cond_expression)`.
+    if (!cond.getType().isa<mlir::IntegerType>()) {
+      auto zeroValue = builder.create<moore::ConstantOp>(loc, type, 0);
+      cond = builder.create<moore::InEqualityOp>(loc, cond, zeroValue);
+    }
```

Comment:
Can you use the `BoolCast` op for this? That should give you an `i1` value more easily.

Reply:
I tried but it didn't work. It seems that the result type `SingleBitType` in `BoolCast` cannot satisfy some constraints of `I1`.

Reply:
Okay, I'll do it this way